GHC 6.2.2 -- mhs: SLIT/FSLIT over boxed String; Pretty's LitString path
HsVersions.h defines SLIT(x) and FSLIT(x) as unboxed string literals
("..."#), a GHC-only syntax MicroHs does not lex; 2,000+ uses across the
compiler. Redefined over the boxed String: FSLIT is a FastString either
way and LitString is a FastString in the shim (ghc622/shims,
utils/FastString). HsVersions.h also hides Semigroup's (<>) and (<$>) from
the Prelude -- Pretty/Outputable/TcHsSyn define their own -- and imports
GhcCompat (common/shims) for the handful of helpers the other patches use.
utils/Pretty.lhs loses the Addr#-based LStr/ptext/hPutLitString fast path
that only existed to serve the unboxed literals; ptext is now ftext.
Same concern as 4.08.2's patch 0002 (DECISIONS.md D45).

--- a/ghc/compiler/HsVersions.h
+++ b/ghc/compiler/HsVersions.h
@@ -82,8 +82,19 @@
 import qualified FastString 
 #endif
 
-#define SLIT(x)	 (FastString.mkLitString# (x#))
-#define FSLIT(x) (FastString.mkFastString# (x#))
+-- GHC 6.0 predates Semigroup, and defines its own (<>) for pretty-printing
+-- documents (Pretty/Outputable). Modern base exports Semigroup's (<>) from the
+-- Prelude, which makes every use in the compiler ambiguous. An explicit
+-- Prelude import suppresses the implicit one, so this hides just that name.
+-- (<$>) likewise: TcHsSyn defines its own, and modern base exports one.
+import Prelude hiding ((<>), (<$>))
+import GhcCompat
+
+-- Both macros expand to an unboxed string literal ("..."#), which MicroHs does
+-- not lex. Applied to a boxed String they denote the same values: FSLIT is a
+-- FastString either way, and LitString is a FastString in the shim.
+#define SLIT(x)	 (FastString.mkLitString x)
+#define FSLIT(x) (FastString.mkFastString x)
 
 #endif /* HsVersions.h */
 
--- a/ghc/compiler/utils/Pretty.lhs
+++ b/ghc/compiler/utils/Pretty.lhs
@@ -178,26 +178,14 @@
 #include "HsVersions.h"
 
 import FastString
-import PrimPacked 	( strLength )
 
 import GLAEXTS
 
 import Numeric (fromRat)
 import IO
 
-#if __GLASGOW_HASKELL__ < 503
-import IOExts		( hPutBufFull )
-#else
-import System.IO	( hPutBuf )
-#endif
 
-#if __GLASGOW_HASKELL__ < 503
-import PrelBase		( unpackCString# )
-#else
-import GHC.Base		( unpackCString# )
-#endif
 
-import PrimPacked	( Ptr(..) )
 
 -- Don't import Util( assertPanic ) because it makes a loop in the module structure
 
@@ -511,7 +499,6 @@
 data TextDetails = Chr  Char
                  | Str  String
                  | PStr FastString	-- a hashed string
-		 | LStr Addr# Int#	-- a '\0'-terminated array of bytes
 
 space_text = Chr ' '
 nl_text    = Chr '\n'
@@ -601,13 +588,11 @@
 char  c = textBeside_ (Chr c) 1# Empty
 text  s = case length   s of {IBOX(sl) -> textBeside_ (Str s)  sl Empty}
 ftext s = case lengthFS s of {IBOX(sl) -> textBeside_ (PStr s) sl Empty}
-ptext (Ptr s) = case strLength (Ptr s) of {IBOX(sl) -> textBeside_ (LStr s sl) sl Empty}
+-- GHC 6.0's ptext takes a LitString: an Addr# to a NUL-terminated C string,
+-- produced by SLIT's unboxed literal. With SLIT redefined over a boxed String
+-- (HsVersions.h) LitString is a FastString in the shim, so this is ftext.
+ptext s = ftext s
 
--- RULE that turns (text "abc") into (ptext (A# "abc"#)) to avoid the
--- intermediate packing/unpacking of the string.
-{-# RULES 
-  "text/str" forall a. text (unpackCString# a) = ptext (Ptr a)
- #-}
 
 nest IBOX(k)  p = mkNest k (reduceDoc p)        -- Externally callable version
 
@@ -922,16 +907,7 @@
 string_txt (Chr c)   s  = c:s
 string_txt (Str s1)  s2 = s1 ++ s2
 string_txt (PStr s1) s2 = unpackFS s1 ++ s2
-string_txt (LStr s1 _) s2 = unpackLitString s1 ++ s2
 
-unpackLitString addr =
- unpack 0#
- where
-    unpack nh
-      | ch `eqChar#` '\0'# = []
-      | otherwise   = C# ch : unpack (nh +# 1#)
-      where
-	ch = indexCharOffAddr# addr nh
 \end{code}
 
 \begin{code}
@@ -1023,20 +999,8 @@
     put (Chr c)  next = hPutChar hdl c >> next 
     put (Str s)  next = hPutStr  hdl s >> next 
     put (PStr s) next = hPutFS   hdl s >> next 
-    put (LStr s l) next = hPutLitString hdl s l >> next 
 
     done = hPutChar hdl '\n'
 
-#if __GLASGOW_HASKELL__ < 503
-hPutBuf = hPutBufFull
-#endif
 
-  -- some versions of hPutBuf will barf if the length is zero
-hPutLitString handle a# 0# = return ()
-hPutLitString handle a# l#
-#if __GLASGOW_HASKELL__ < 411
-  = hPutBuf handle (A# a#) (I# l#)
-#else
-  = hPutBuf handle (Ptr a#) (I# l#)
-#endif
 \end{code}
